README
Hook::Filter version 0.01
=========================

Hook::Filter is a runtime firewall for subroutine calls.

See perldoc Hook::Filter for more information.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Module::Pluggable
  File::Spec

AUTHOR

Erwan Lemonnier

COPYRIGHT AND LICENCE

See perldoc Hook::Filter




Changes
Revision history for Perl extension Hook::Filter

0.01  2006-03-09
	- first working version

0.02  2006-05-23
        - adding some tests

0.03  2007-05-16
        - moved repository to sourceforge (https://sourceforge.net/projects/hook-filter)
        - removed the tests from_pkg(), from_sub() and is_sub(), hence
          breaking the backward compatibility of filtering rules
        - added the tests from(), arg() and subname()
        - allowed injection of new rules during runtime
        - updated POD

0.04  2007-05-22
        - more POD
        - using a rule file is now optional

0.05  2007-05-23
        - allow multiple use of 'rules' if the file stays the same

0.06  2007-05-23
        - POD updates

0.07  2007-05-24
        - refactored to avoid 'Too late to run INIT' block message when used from eval/require
        - Hook::Filter now *must* be used with parameter 'hook =>'

0.08  2007-05-25
        - better workaround for skipped INIT block when evaled/required
        - handle warnings in tests

0.09  2008-06-09
        - kvalite++
        - fix test


lib/Hook/Filter/Hook.pm
#################################################################
#
#   Hook::Filter::Hook - Wrap subroutines in a filtering closure
#
#   $Id: Hook.pm,v 1.1 2006/01/27 06:35:48 erwan Exp $
#
#   060302 erwan Created
#

package Hook::Filter::Hook;

use 5.006;
use strict;
use warnings;
use Carp qw(croak);
use Data::Dumper;
use Symbol;
use base qw(Exporter);

our @EXPORT = qw(get_caller_package
		 get_caller_file
		 get_caller_line
		 get_caller_subname
		 get_subname
		 get_arguments
		 );

our $VERSION = '0.01';

use vars qw($CALLER_PACKAGE $CALLER_FILE $CALLER_LINE $CALLER_SUBNAME $SUBNAME @ARGUMENTS);

#----------------------------------------------------------------
#
#
#   CLASS FUNCTIONS
#
#
#----------------------------------------------------------------

#----------------------------------------------------------------
#
#   accessors for use in Hook::Filter::Plugins:: modules
#

sub get_caller_package   { return $CALLER_PACKAGE; };
sub get_caller_file      { return $CALLER_FILE; };
sub get_caller_line      { return $CALLER_LINE; };
sub get_caller_subname   { return $CALLER_SUBNAME; };
sub get_subname          { return $SUBNAME; };
sub get_arguments        { return @ARGUMENTS; };

#----------------------------------------------------------------
#
#
#   OBJECT METHODS
#
#
#----------------------------------------------------------------

#----------------------------------------------------------------
#
#   new - build a new Hook::Filter::Hook
#

sub new {
    my($pkg) = @_;
    $pkg = ref $pkg || $pkg;
    my $self = bless({},$pkg);
    
    $self->{RULES} = [];
    $self->{SUBS} = {};
    
    return $self;
}

#----------------------------------------------------------------
#
#   register_rule - register a new Hook::Filter::Rule into self
#

sub register_rule {
    my($self,$rule) = @_;
    
    if (!defined $rule || ref $rule ne "Hook::Filter::Rule" || scalar @_ != 2) {
	shift @_;
	croak "invalid parameter: Hook::Filter::Hook->register_rule expects one instance of Hook::Filter::Rule, and not [".Dumper(@_)."]";
    }
    
    # TODO: silently skip registering the same rule twice?
    push @{$self->{RULES}}, $rule;
    
    return $self;
}

#----------------------------------------------------------------
#
#   flush_rules - erase all known rules
#

sub flush_rules {
    my($self) = @_;

    $self->{RULES} = [];

    return $self;
}

#----------------------------------------------------------------
#
#   filter_sub - build a filter closure wrapping calls to the provided sub
#

sub filter_sub {
    my($self,$subname) = @_;
    
    if (!defined $subname || ref \$subname ne "SCALAR" || scalar @_ != 2) {
	shift @_;
	croak "invalid parameter: Hook::Filter::Hook->filter_sub expects a subroutine name, but got [".Dumper(@_)."].";
    }

    if ($subname !~ /^(.+)::([^:]+)$/) {
	croak "invalid parameter: [$subname] is not a valid subroutine name (must include package name).";
    }

    my($pkg,$func) = ($1,$2);

    # check whether subroutine is already filtered, and skip if so
    if (exists $self->{SUBS}->{$subname}) {
	return $self;
    }
    
    my $filtered_func = *{ qualify_to_ref($func,$pkg) }{CODE};

    # create the closure that will replace $func in package $pkg
    my $filter = sub {
	my(@args) = @_;
	
	# TODO: looking at source for Hook::WrapSub, it might be a good idea to copy/paste some of its code here, to build a valid caller stack
	# TODO: look at Hook::LexWrap and fix so that caller() work in subroutines
	
	# set global variables
	$CALLER_PACKAGE  = (caller(0))[0];
	$CALLER_FILE     = (caller(0))[1];
	$CALLER_LINE     = (caller(0))[2];
	$CALLER_SUBNAME  = (caller(1))[3] || "";
	$SUBNAME         = $subname;
	@ARGUMENTS       = @args;

	# evaluate all rules, until all one is found to be true or all are found to be false
	foreach my $rule (@{$self->{RULES}}) {
	    if ($rule->eval()) {
		# found one rule that is true -> call subroutine in right context
		if (wantarray) {
		    my @results = $filtered_func->(@args);
		    return @results;
		} else {
		    my $result = $filtered_func->(@args);
		    return $result;
		}
	    }
	}

	# all rules evaluated to false, so skip subroutine, and fake a return value (bleh.)
	if (wantarray) {
	    return ();
	}
	return;
    };

    # keep track of already hooked subroutines
    $self->{SUBS}->{$subname} = 1;
    
    # replace $package::$func with our closure
    no strict 'refs';    
    no warnings;

    *{ qualify_to_ref($func,$pkg) } = $filter;

    return $self;
}

1;

__END__

=head1 NAME

Hook::Filter::Hook - Wrap subroutines in a filtering closure

=head1 VERSION

$Id: Hook.pm,v 1.1 2006/01/27 06:35:48 erwan Exp $

=head1 DESCRIPTION

=head1 SYNOPSIS

    use Hook::Filter::Hook;
    use Hook::Filter::Rule;

    my $hook = new Hook::Filter::Hook();

    my $rule = new Hook::Filter::Rule("in_package('My::Package') && in_sub('my_sub')");
    $hook->register_rule($rule);

    $hook->filter_sub("My::Package","mylog");

Later on, you may want to change the rules during runtime:

    $hook->flush_rules();
    $hook->register_rule($newrule1);
    $hook->register_rule($newrule2);

=head1 INTERFACE - METHODS

=over 4

=item B<new>()

Return a new generic Hook::Filter::Hook instance.

=item B<register_rule>(I<$rule>)

Add a new Hook::Filter::Rule I<$rule> to this Hook::Filter::Hook instance.

=item B<filter_sub>(I<$subname>)

Filter the function or method I<$subname>. I<$subname> must be of the form
C<< package_name::function_name >>.
All calls to C<< $subname >> will thereafter be redirected
to a wrapper closure that will evaluate all the rules registered in
Hook::Filter::Hook and if one of the rules evals to true,
the original function C<< $package::$sub() >> will be called normally.
Otherwise it will be skipped and its result simulated.

=item B<flush_rules>()

Remove all rules currently registered in this Hook::Filter::Hook.

=back

=head1 INTERFACE - CLASS FUNCTIONS

The following class functions are to be used by modules under
Hook::Filter::Plugins that implement specific test functions
for use in filter rules. 

Any use of the following functions in a different context than
inside the implementation of a filter rule test is guaranteed
to return only garbage. 

See Hook::Filter::Plugins::Location for a usage example.

=over 4

=item B<get_caller_package()> 

Return the name of the package calling the filtered subroutine.

=item B<get_caller_file()>

Return the name of the file calling the filtered subroutine.

=item B<get_caller_line()> 

Return the line number at which the filtered subroutine was called.

=item B<get_caller_subname()> 

Return the complete name (package+name) of the subroutine calling the filtered subroutine. 
If the subroutine was called directly from the main namespace, return an empty string.

=item B<get_subname()> 

Return the complete name of the filtered subroutine for which the rules
are being eval-ed.

=item B<get_arguments()>

Return the list of arguments that would be passed to the filtered subroutine.

=back

=head1 DIAGNOSTICS

=over 4

=item C<< $hook->register_rule($rule) >> croaks if I<$rule> is not a Hook::Filter::Rule.

=item C<< $hook->filter_sub(I<$pkg>,I<$func>) >> croaks when passed invalid arguments.

=item The closure wrapping all filtered subroutines emits perl warning when rules die upon being eval-ed.

=back

=head1 BUGS AND LIMITATIONS

See Hook::Filter

=head1 SEE ALSO

See Hook::Filter, Hook::Filter::Rule.

=head1 AUTHOR

Erwan Lemonnier C<< <erwan@cpan.org> >>.

=head1 COPYRIGHT AND LICENSE

See Hook::Filter.

=head1 DISCLAIMER OF WARRANTY

See Hook::Filter.

=cut





lib/Hook/Filter/Rule.pm
#################################################################
#
#   Hook::Filter::Rule - A filter rule
#
#   $Id: Rule.pm,v 1.6 2007/05/24 14:52:37 erwan_lemonnier Exp $
#
#   060301 erwan Created
#   070516 erwan Small POD and layout fixes
#   070524 erwan Used BEGIN instead of INIT
#

package Hook::Filter::Rule;

use 5.006;
use strict;
use warnings;
use Carp qw(croak);
use Data::Dumper;
use Symbol;
use Module::Pluggable search_path => ['Hook::Filter::Plugins'], require => 1;

our $VERSION='0.03';

#----------------------------------------------------------------
#
#   load test functions from plugins
#

BEGIN {

    my %TESTS;

    foreach my $plugin (Hook::Filter::Rule->plugins()) {
	my @tests = $plugin->register();
	# TODO: test that @tests is an array of strings. die with BUG:

	foreach my $test ($plugin->register()) {
	    if (exists $TESTS{$test}) {
		croak "invalid plugin function: test function [$test] exported by plugin [$plugin] is already exported by an other plugin.";
	    }
	    *{ qualify_to_ref($test,"Hook::Filter::Rule") } = *{ qualify_to_ref($test,$plugin) };
	    $TESTS{$test} = 1;
	}
    }
}

#----------------------------------------------------------------
#
#   new - build a new filter rule
#

sub new {
    my($pkg,$rule) = @_;
    $pkg = ref $pkg || $pkg;
    my $self = bless({},$pkg);

    if (!defined $rule || ref \$rule ne "SCALAR" || scalar @_ != 2) {
	shift @_;
	croak "invalid parameter: Hook::Filter::Rule->new expects one string describing a filter rule, but got [".Dumper(@_)."].";
    }

    $self->{RULE} = $rule;

    return $self;
}

#----------------------------------------------------------------
#
#   rule - accessor for the rule
#

sub rule {
    return $_[0]->{RULE};
}

#----------------------------------------------------------------
#
#   source - where the rule came from (used in error messages only)
#

sub source {
    my($self,$orig) = @_;

    if (!defined $orig || ref \$orig ne "SCALAR" || scalar @_ != 2) {
	shift @_;
	croak "invalid parameter: Hook::Filter::Rule->source expects one string, but got [".Dumper(@_)."].";
    }

    $self->{SOURCE} = $orig;
}

#----------------------------------------------------------------
#
#   eval - evaluate a rule. return either true or false
#

sub eval {
    my $self = shift;
    my $rule = $self->{RULE};

    my $res = eval $rule;
    if ($@) {
	# in doubt, let's assume we are not filtering anything, ie allow function calls as if we were not here
	warn "WARNING: invalid Hook::Filter rule [$rule] ".
	    ( (defined $self->{SOURCE})?"from file [".$self->{SOURCE}."] ":"")."caused error:\n".
	    "[".$@."]. Assuming this rule returned true.\n";
	return 1;
    }

    return ($res)?1:0;
}

1;

__END__

=head1 NAME

Hook::Filter::Rule - A hook filter rule

=head1 DESCRIPTION

A filter rule is a string containing a perl expression that evaluates to
either true or false.

A rule may contain calls to functions exported by any module under
C<< Hook::Filter::Plugins:: >>.

=head1 SYNOPSIS

    use Hook::Filter::Rule;

    my $rule = Hook::Filter::Rule->new("1");
    if ($rule->eval) {
	print "just now, the rule [".$rule->rule."] is true\n";
    }

=head1 INTERFACE

=over 4

=item C<< my $r = new($rule) >>

Return a new C<Hook::Filter::Rule> created from the string C<$rule>. C<$rule>
is a valid line of perl code that should return either true or false when
eval-ed. It can contain calls to any of the functions exported by the plugin modules
located under C<< Hook::Filter::Plugins:: >>.

=item C<< $r->eval() >>

Eval this rule. Return 0 if the rule eval-ed to false. Return 1 if the rule eval-ed
to true, or if the rule died/croaked.

If the rule dies/croaks/confesses while being eval-ed, a perl warning is
thrown and the rule is assumed to return true (fail-safe). The warning
contains details about the error message, the rule itself and where it
comes from (as specified with C<< source() >>).

=item C<< $r->source($message) >>

Specify the origin of this rule. If the rule was parsed from a rule file,
C<$message> should be the path to this file. This is used in the warning
message emitted when a rule dies during C<< eval() >>.

=item C<< $r->rule() >>

Return the rule's string (C<$rule> in C<< new() >>).

=back

=head1 DIAGNOSTICS

=over 4

=item C<< use Hook::Filter::Rule >> croaks if a plugin module tries to export a function name
that is already exported by an other plugin.

=item C<< Hook::Filter::Rule->new($rule) >> croaks if C<$rule> is not a scalar.

=item C<< $rule->eval() >> will emit a perl warning if the rule dies when eval-ed.

=item C<< $rule->source($text) >> croaks if C<$text> is not a scalar.

=back

=head1 BUGS AND LIMITATIONS

See Hook::Filter

=head1 SEE ALSO

See Hook::Filter, Hook::Filter::RulePool, Hook::Filter::Hooker, Hook::Filter::Plugins::Library.

=head1 VERSION

$Id: Rule.pm,v 1.6 2007/05/24 14:52:37 erwan_lemonnier Exp $

=head1 AUTHOR

Erwan Lemonnier C<< <erwan@cpan.org> >>

=head1 LICENSE

See Hook::Filter.

=cut


lib/Hook/Filter/Hooker.pm
#################################################################
#
#   Hook::Filter::Hooker - Wrap subroutines in a firewalling closure
#
#   $Id: Hooker.pm,v 1.5 2007/05/23 08:26:15 erwan_lemonnier Exp $
#
#   060302 erwan Created
#   070516 erwan Use the rule pool
#

package Hook::Filter::Hooker;

use strict;
use warnings;
use Carp qw(croak);
use Data::Dumper;
use Symbol;
use base qw(Exporter);
use Hook::Filter::RulePool qw(get_rule_pool);

our @EXPORT = qw();
our @EXPORT_OK = qw( get_caller_package
		     get_caller_file
		     get_caller_line
		     get_caller_subname
		     get_subname
		     get_arguments
		     filter_sub
		     );



use vars qw( $CALLER_PACKAGE
	     $CALLER_FILE
	     $CALLER_LINE
	     $CALLER_SUBNAME
	     $SUBNAME
	     @ARGUMENTS );

# singleton instance of Hook::Filter::RulePool
my $pool = get_rule_pool();

# a hash whose keys are the fully qualified names of all filtered
# subroutines, to avoid filtering one twice
my %subs;

#----------------------------------------------------------------
#
#   accessors for use in Hook::Filter::Plugins:: modules
#

sub get_caller_package   { return $CALLER_PACKAGE; };
sub get_caller_file      { return $CALLER_FILE;    };
sub get_caller_line      { return $CALLER_LINE;    };
sub get_caller_subname   { return $CALLER_SUBNAME; };
sub get_subname          { return $SUBNAME;        };
sub get_arguments        { return @ARGUMENTS;      };

#----------------------------------------------------------------
#
#   filter_sub - build a filter closure wrapping calls to the provided sub
#

sub filter_sub {
    my $subname = shift;

    if (!defined $subname || ref \$subname ne "SCALAR" || scalar @_) {
	shift @_;
	croak "invalid parameter: Hook::Filter::Hooker->filter_sub expects a subroutine name, but got [".Dumper($subname,@_)."].";
    }

    if ($subname !~ /^(.+)::([^:]+)$/) {
	croak "invalid parameter: [$subname] is not a valid subroutine name (must include package name).";
    }

    my ($pkg,$func) = ($1,$2);

    # check whether subroutine is already filtered, and skip if so
    return if (exists $subs{$subname});

    my $filtered_func = *{ qualify_to_ref($func,$pkg) }{CODE};

    # create the closure that will replace $func in package $pkg
    my $filter = sub {
	my (@args) = @_;

	# TODO: looking at source for Hook::WrapSub, it might be a good idea to copy/paste some of its code here, to build a valid caller stack
	# TODO: look at Hook::LexWrap and fix so that caller() work in subroutines

	# set global variables
	$CALLER_PACKAGE  = (caller(0))[0];
	$CALLER_FILE     = (caller(0))[1];
	$CALLER_LINE     = (caller(0))[2];
	$CALLER_SUBNAME  = (caller(1))[3] || "";
	$SUBNAME         = $subname;
	@ARGUMENTS       = @args;

	# evaluate all rules. if true is returned, forward the call
	if ($pool->eval_rules) {
	    if (wantarray) {
		my @results = $filtered_func->(@args);
		return @results;
	    } else {
		my $result = $filtered_func->(@args);
		return $result;
	    }
	}

	# the call was blocked. fake a return value (ugly.)
	if (wantarray) {
	    return ();
	}
	return;
    };

    # keep track of already hooked subroutines
    $subs{$subname} = 1;

    # replace $package::$func with our closure
    no strict 'refs';
    no warnings;

    *{ qualify_to_ref($func,$pkg) } = $filter;
}

1;

__END__

=head1 NAME

Hook::Filter::Hooker - Wrap subroutines in a firewalling closure

=head1 DESCRIPTION

This module is used internaly by Hook::Filter to generate an anonymous
sub that is wrapped around each filtered subroutine and either forwards
the call to the subroutine or block it and spoofs return values (undef or
an empty list depending on context).

=head1 SYNOPSIS

    use Hook::Filter::Hooker;

    my $hooker = new Hook::Filter::Hooker();

    $hooker->filter_sub("My::Package");

    # mylog is declared in the current package
    $hooker->filter_sub("mylog");

=head1 INTERFACE

C<Hook::Filter::Hooker> exports no functions by default. But the following functions
can be explicitly imported upon using C<Hook::Filter::Hooker>:

=over 4

=item C<< $hooker->filter_sub($subname) >>

Add a filter around the subroutine C<$subname>. I<$subname> must either be a fully qualified
function name, or the name of a function located in the current package.

All calls to C<< $subname >> will thereafter be redirected
to a wrapper closure that will evaluate all the rules registered in
C<Hook::Filter::RulePool> using the method C<eval()> on the pool.
If C<eval()> returns true, the call is forwarded, otherwise it is
blocked.

=back

The following class functions are to be used by modules under
C<Hook::Filter::Plugins::> that implement specific test functions
for use in filter rules.

Any use of these functions in a different context than
inside a plugin implementation is guaranteed
to return only garbage.

See C<Hook::Filter::Plugins::Library> for a usage example.

=over 4

=item C<get_caller_package()>

Return the name of the package calling the filtered subroutine.

=item C<get_caller_file()>

Return the name of the file calling the filtered subroutine.

=item C<get_caller_line()>

Return the line number at which the filtered subroutine was called.

=item C<get_caller_subname()>

Return the complete name (package+name) of the subroutine calling the filtered subroutine.
If the subroutine was called directly from the main namespace, return an empty string.

=item C<get_subname()>

Return the complete name of the filtered subroutine for which the rules
are being eval-ed.

=item C<get_arguments()>

Return the list of arguments that would be passed to the filtered subroutine.

=back

=head1 DIAGNOSTICS

=over 4

=item C<< $hook->filter_sub(I<$pkg>,I<$func>) >> croaks when passed invalid arguments.

=item The closure wrapping all filtered subroutines emits a perl warning when a rule dies upon being eval-ed.

=back

=head1 BUGS AND LIMITATIONS

See Hook::Filter

=head1 SEE ALSO

See Hook::Filter, Hook::Filter::Rule, Hook::Filter::RulePool, Hook::Filter::Plugins::Library.

=head1 VERSION

$Id: Hooker.pm,v 1.5 2007/05/23 08:26:15 erwan_lemonnier Exp $

=head1 AUTHOR

Erwan Lemonnier C<< <erwan@cpan.org> >>.

=head1 LICENSE

See Hook::Filter.

=cut





lib/Hook/Filter.pm
#################################################################
#
#   Hook::Filter - A runtime filtering layer on top of subroutine calls
#
#   $Id: Filter.pm,v 1.11 2008/06/09 21:04:38 erwan_lemonnier Exp $
#
#   051105 erwan Created
#   060301 erwan Recreated
#   070516 erwan Updated POD and license, added flush_rules and add_rule
#   070522 erwan More POD + don't use rule file unless 'rules' specified in import
#   070523 erwan Can use 'rules' multiple time if same rule file specified
#   070523 erwan POD updates
#   070524 erwan Import parameter 'hook' is now mandatory
#

package Hook::Filter;

use 5.006;
use strict;
use warnings;
use Carp qw(confess croak);
use File::Spec;
use Hook::Filter::Rule;
use Hook::Filter::RulePool qw(get_rule_pool);
use Hook::Filter::Hooker qw(filter_sub);
use base qw(Exporter);
use Data::Dumper;

our @EXPORT = qw();

our $VERSION = '0.09';

#----------------------------------------------------------------
#
#   Global vars
#
#----------------------------------------------------------------

# the rule file actually used by Hook::Filter, and as declared with parameter 'rules'
my $RULES_FILE;

# list of subs to hijack
my %HOOK_SUBS;

sub _queue_sub {
    my ($pkg,$name) = @_;
    ($name =~ /::/) ? $HOOK_SUBS{$name}=1 : $HOOK_SUBS{$pkg."::".$name}=1;
}

#----------------------------------------------------------------
#
#   import - verify import parameters, filter the subs and load the rule file
#

sub import {
    my($class,%args) = @_;
    my $pkg = caller(0);

    #
    # check parameter 'rules', indicating path to the rule file
    #

    if (exists $args{rules}) {

	croak "import parameter 'rules' for Hook::Filter should be a string, but was undef."
	    if (!defined $args{rules});

	croak "import parameter 'rules' for Hook::Filter should be a string, but was [".Dumper($args{rules})."]."
	    if (ref \$args{rules} ne 'SCALAR');

	croak "you tried to specify 2 different Hook::Filter rule file: [$RULES_FILE] and [".$args{rules}."]. you may have only 1 rule file."
	    if (defined $RULES_FILE && $RULES_FILE ne $args{rules});

	$RULES_FILE = $args{rules};
	delete $args{rules};

	#
	# load the rule file, if any
	#

	my $pool = get_rule_pool();

	if (-f $RULES_FILE) {
	    # TODO: support runtime monitoring of rules file and update of rules upon changes in file

	    open(IN,"$RULES_FILE")
		or confess "failed to open Hook::Filter rules file [$RULES_FILE]: $!";
	    while (my $line = <IN>) {
		chomp $line;
		next if ($line =~ /^\s*\#/);
		next if ($line =~ /^\s*$/);

		my $rule = new Hook::Filter::Rule($line);
		$rule->source($RULES_FILE);
		$pool->add_rule($rule);
	    }
	    close(IN);
	}
    }

    #
    # check parameter 'hook', indicating which subroutines to filter
    #

    croak "you must call Hook::Filter with the import parameter 'hook' set to something"
	if (!exists $args{hook});

    croak "Invalid parameter: 'hook' should be a string or an array of strings, but was undef."
	if (!defined $args{hook});

    if (ref $args{hook} eq 'ARRAY') {
	foreach my $name (@{$args{hook}}) {
	    if (ref \$name ne 'SCALAR') {
		croak "Invalid parameter: 'hook' for Hook::Filter should be a string or an array of strings, but was [".Dumper($args{hook})."].";
	    }
	    _queue_sub($pkg,$name);
	}
    } elsif (ref \$args{hook} eq 'SCALAR') {
	_queue_sub($pkg,$args{hook});
    } else {
	croak "Invalid parameter: 'hook' for Hook::Filter should be a string or an array of strings, but was [".Dumper($args{hook})."].";
    }

    delete $args{hook};

    # propagate super class's import
    $class->export_to_level(1,undef,());
}

#
# when all is compiled, do filter all the subs
#

sub _filter_subs {
    map { filter_sub($_) } keys %HOOK_SUBS;
}

# this init block won't be executed if Hook::Filter is used from an eval/require
INIT {
    # add a filtering closure around each sub
    _filter_subs;
}

1;

__END__

=head1 NAME

Hook::Filter - A runtime firewall for subroutine calls

=head1 DESCRIPTION

Hook::Filter is a runtime firewall for subroutine calls.

Hook::Filter lets you wrap one or more subroutines with a filter that
either forwards calls to the subroutine or blocks them, depending on
a number of rules that you define yourself. Each rule is simply one
line of Perl code that must evaluate to false (block the call) or true
(allow it).

The filtering rules are fetched from a file, called the rules file, or
they can be injected dynamically at runtime.

Each time a call is made to one of the filtered subroutines, all the
filtering rules are eval-ed, and if one of them returns true, the
call is forwarded, otherwise it is blocked. If no rules are defined,
all calls are forwarded by default.

Filtering rules are very flexible. You can block or allow calls to
a subroutine based on things such as the caller's identity, the
values of the arguments passed to the subroutine, the structure
of the call stack, or basically any other test that can be implemented
in Perl.

=head1 SYNOPSIS

To filter calls to the local subroutines C<mydebug>, C<myinfo> and
to C<Some::Other::Module::mywarn>:

    use Hook::Filter hook => [ "mydebug" ,"myinfo", "Some::Other::Module::mywarn" ];

To filter calls to the local subroutine C<_debug>, and import filtering
rules from the file C<~/debug.rules>:

    use Hook::Filter hook => '_debug', rules => '~/debug.rules';

The rule file C<~/debug.rules> could contain the following rules:

    # allow calls to 'mydebug' from within module 'My::Filthy:Attempt'
    subname eq 'mydebug' && from =~ /^My::Filthy::Attempt/

    # allow calls only from within a specific subroutine
    from eq 'My::Filthy::Attempt::func'

    # allow calls only if the subroutine's 2nd argument matches /bob/
    args(1) =~ /bob/

    # all other calls to 'myinfo', 'mydebug' or 'mywarn' will be skipped

You could also inject those rules dynamically at runtime:

    use Hook::Filter::RulePool qw(get_rule_pool);

    get_rule_pool->add_rule("subname eq 'mydebug' && from =~ /^My::Filthy::Attempt/");
                 ->add_rule("from =~ /^My::Filthy::Attempt::func$/");
                 ->add_rule("args(1) =~ /bob/");

To see which test functions can be used in rules, see Hook::Filter::Plugins::Library.

=head1 RULES

=head2 SYNTAX

A rule is a string containing one line of valid perl code that returns
either true or false when eval-ed. This line of code is usually made of
boolean operators combining functions that are exported by the modules
located under C<Hook::Filter::Plugins::>. See those modules for more details.

If you specify a rule file with the import parameter C<rules>, the rules
will be parsed out of this file according to the following syntax:

=over 4

=item * any line starting with C<< # >> is a comment.

=item * any empty line is ignored.

=item * any other line is considered to be a rule, ie a valid line of perl code that can be eval-ed.

=back

Each time one of the filtered subroutines is called, all loaded rules
are eval-ed until one returns true or all returned false. If one returns
true, the call is forwarded to the filtered subroutine, otherwise it is
skipped and a return value spoofed: either undef or an empty list,
depending on the context.

If a rule dies/croaks/confess upon being eval-ed (f.ex. when you left
a syntax error in the rule's string), it will be assumed
to have returned true. This is a form of fail-safe policy. You will
also get a warning message with a complete diagnostic.

=head2 RULE POOL

All rules are stored in a rule pool. You can use this pool to access
and manipulate rules during runtime.

There are 2 mechanisms to load rules into the pool:

=over 4

=item * Rules can be imported from a file at INIT time. Just specify the path
and name of this file with the import parameter C<< rules >>, and fill
this file with rules as shown in SYNOPSIS.

=item * Rules can also be injected dynamically at runtime. The following code
injects a rule that is always true, hence always allowing calls to the
filtered subroutines:

    use Hook::Filter::RulePool qw(get_rule_pool);
    get_rule_pool->add_rule("1");

=back

Rules can all be flushed at runtime:

    get_rule_pool->flush_rules();

For other operations on rules, see the modules C<Hook::Filter::RulePool>
and C<Hook::Filter::Rule>.

=head2 PASS ALL CALLS BY DEFAULT

If no rules are registered in the rule pool, or if all registered rules
die/croak when eval-ed, the default behaviour is to allow all calls to
the filtered subroutines.

That would happen for example if you specify no rule file via the import
parameter C<rules> and register no rules dynamically afterward.

To change this default behaviour, just add one default rule
that always returns false:

    use Hook::Filter::RulePool qw(get_rule_pool);
    get_rule_pool->add_rule("0");

All calls to the filtered subroutines are then blocked by default, as
long as no rule evals to true.

=head2 EXTENDING THE PLUGIN LIBRARY

The default plugin C<Hook::Filter::Plugins::Library> offers a number of
functions that can be used inside the filter rules, but you may want
to extend this library with your own functions.

You can easily do that by writing a new plugin module having the same
structure as C<Hook::Filter::Plugins::Library> and placing it under
C<Hook/Filter/Plugins/>. See C<Hook::Filter::Hooker> and
C<Hook::Filter::Plugins::Library> for details on how to do that.

=head1 INTERFACE

C<Hook::Filter> exports no  functions, but C<Hook::Filter> accepts the following
import parameters:

=over 4

=item C<< rules => $rules_file >>

Optional. Specify the complete path to a rule file. This import parameter can be used
only once in a program (usually in package C<< main >>) independently of how many
times C<< Hook::Filter >> is used. The file is parsed at INIT time.

See the RULES section for details.

Example:

    # look for rules in the local file 'my_rules'
    use Hook::Filter rules => 'my_rules';

=item C<< hook => $subname1 >> or C<< hook => [$subname1,$subname2...] >>

Mandatory. Specify which subroutines to filter. C<$subname> can either be a fully
qualified name or just the name of a subroutine located in the current
package.

Examples:

    # filter function debug() in the current package
    use Hook::Filter hook => 'debug';

    # filter function debug() in an other package
    use Hook::Filter hook=> 'Other::Package::debug';

    # do both at once
    use Hook::Filter hook=> [ 'Other::Package::debug', 'debug' ];

=back

=head1 DIAGNOSTICS

=over 4

=item Passing wrong arguments to C<Hook::Filter>'s import parameters will
cause it to croak.

=item The import parameter C<< hook >> must be used at least once otherwise
C<Hook::Filter> croaks with an error message.

=item An IO error when opening the rule file causes Hook::Filter to die.

=item An error in a filter rule will be reported with a perl warning.

=back

=head1 RESTRICTIONS

=head2 SECURITY

C<Hook::Filter> gives anybody with write permissions toward the rule file
the possibility to inject code into your application. This can be highly
dangerous! Protect your filesystem.

=head2 CAVEATS

=over 4

=item * Return values: when a call to a subroutine is allowed, the input and output arguments
of the subroutine are forwarded without modification. But when the call
is blocked, the subroutine's return value is simulated and will be C<< undef >>
in SCALAR context and an empty list in ARRAY context. Therefore, DO NOT filter
subroutines whose return values are significant for the rest of your code.

=item * Speed: Hook::Filter evaluates all filter rules for each call to a
filtered subroutine, which is slow. It would therefore be very unappropriate
to filter a heavily used subroutine in speed requiring applications.

=back

=head2 THREADS

Hook::Filter is not thread safe.

=head2 KEEP IT SIMPLE

The concept of blocking/allowing subroutine calls dynamically is somewhat
unusual and fun. Don't let yourself get too excited though. Doing that kind of
dynamic stuff makes your code harder to understand for non-dynamic developers,
hence reducing code stability.

=head2 USING Hook::Filter VIA REQUIRE/EVAL

If you do something like:

    eval "use Hook::Filter hook => 'some_sub'";

You will get a 'Too late to run INIT block' warning, and the subroutine
C<some_sub> will not be filtered.

There is unfortunately no simple way to fix that.

A rather ugly work-around would be to run explicitly the private
function C<_filter_subs> from C<Hook::Filter>:

    {
        no warnings 'void';
        eval "use Hook::Filter hook => 'some_sub', qw(filter_subs)";
    }
    ...

    # later on, call filter_subs explicitly
    Hook::Filter::_filter_subs;

=head1 USE CASE

Why would one need a firewall for subroutine calls?
Here are a couple of relevant use cases:

=over 4

=item * A large application logs a lot of information. You want to implement
a logging policy to limit the amount of logged information, but you don't want
to modify the logging code. You do that by filtering the functions defined in
the logging API with C<Hook::Filter>, and by defining a rule file that implements
your logging policy.

=item * A large application crashes regularly so you decide to turn on debugging
messages system wide with full verbosity. You get megazillions of log messages.
Instead of greping your way through them or starting your debugger, you use C<Hook::Filter>
to filter the function that logs debug messages and define tailored rules that
allow only relevant debug messages to be logged.

=back

=head1 SEE ALSO

See Hook::Filter::Rule, Hook::Filter::RulePool, Hook::Filter::Plugins::Library, Hook::Filter::Hooker.
See even Hook::WrapSub, Log::Localized, Log::Log4perl, Log::Dispatch.

=head1 BUGS AND LIMITATIONS

Please report any bugs or feature requests to
C<bug-hook-filter@rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org>.

=head1 REPOSITORY

The source of Hook::Filter is hosted at sourceforge. You can access
it at https://sourceforge.net/projects/hook-filter/.

=head1 AUTHOR

Written by Erwan Lemonnier C<< <erwan@cpan.org> >> based on inspiration
received during the 2005 Nordic Perl Workshop. Kind thanks to Claes Jakobsson &
Jerker Montelius for their suggestions and support!

=head1 LICENSE

See the LICENSE file included in this distribution.

=cut






